modules/indigo/module/inc/indigo/of_state_manager.h

```diff
@@ -141,13 +141,34 @@ indigo_core_stats_get(uint32_t *total_flows,
  * called when a gentable_entry_add message is received for a key that
  * already exists in the table.
  *
+ * Async operation support:
+ * add4()/modify4()/del4() are used for async operations.
+ * These new APIs can return INDIGO_ERROR_PENDING to state manager
+ * if the operations are async.
+ * The state manager provides a resume function,
+ * indigo_core_gentable_resume(), to complete the remaining works.
+ * The state manager also needs the operation context, containing
+ * connection id and oflow obj, to resume its remaining tasks.
+ * The operation context is passed to the driver in a form of
+ * struct indigo_core_op_context. (See add4i()/modify4()/del4()).
+ * Drivers call the resume function with the operation context and
+ * the final status.
  ****************************************************************/
 
+
 /**
  * @brief Opaque handle to a gentable
  */
 typedef struct indigo_core_gentable indigo_core_gentable_t;
 
+/**
+ * @brief Opaque handle to an operation
```

Comment:
my understanding is that you intend that indigo_core_op_context_t to be opaque.  but it's typedefed here and is not preceded by sufficiently strong language to tell the user they are not intended to use its contents.  my suggestion is to move the typedef to ofstatemanager_int.h and replace ```indigo_core_op_context_t*``` with ```void*```, and then cast it to ```indigo_core_op_context_t*``` internally.

Reply:
This is the solution after discussion. We cannot make it to be opaque since some gentables need the connection Id to send async event to controller.